bluemira/codes/openmc/csg_reactor.py

```diff
@@ -0,0 +1,466 @@
+# SPDX-FileCopyrightText: 2021-present M. Coleman, J. Cook, F. Franza
+# SPDX-FileCopyrightText: 2021-present I.A. Maione, S. McIntosh
+# SPDX-FileCopyrightText: 2021-present J. Morris, D. Short
+#
+# SPDX-License-Identifier: LGPL-2.1-or-later
+"""Putting together csg components to form a csg reactor model in the OpenMC universe."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from itertools import chain
+from typing import TYPE_CHECKING
+
+import numpy as np
+import openmc
+import openmc.region
+
+from bluemira.codes.openmc.make_cell import BlanketCellArray, DivertorCellArray
+from bluemira.codes.openmc.material import CellType
+from bluemira.codes.openmc.tools import (
+    OpenMCEnvironment,
+    flat_intersection,
+    flat_union,
+    round_up_next_openmc_ids,
+)
+from bluemira.geometry.constants import D_TOLERANCE
+from bluemira.geometry.error import GeometryError
+from bluemira.geometry.tools import (
+    is_convex,
+    polygon_revolve_signed_volume,
+)
+from bluemira.radiation_transport.neutronics.constants import to_cm, to_cm3
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
+    from bluemira.codes.openmc.material import MaterialsLibrary
+    from bluemira.radiation_transport.neutronics.neutronics_axisymmetric import (
+        NeutronicsReactor,
+    )
+
+
+@dataclass
+class CSGComponents:
+    """A container for each component represented in CSG geometry."""
+
+    blanket: BlanketCellArray
```

Comment:
Thoughts about naming here for generic situations. Main chamber might be a better name because here we are implicitly saying that if its not divertor its blanket (which is fine here in codes). If its user facing we should probably be a bit more explicit